File: geekbang/algorithm/source-code/2/lc111 二叉树的最小深度.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        # 不同于maxDepth的另一种写法
        # 也可以采用maxDepth的写法
        self.depth = 0
        self.ans = 1 << 30
        self.recur(root)
        return self.ans

    def recur(self, node):
        if not node:
            self.ans = min(self.ans, self.depth)
            return
        if not node.left and not node.right:
            self.ans = min(self.ans, self.depth + 1)
            return
        self.depth += 1
        if node.left:
            self.recur(node.left)
        if node.right:
            self.recur(node.right)
        self.depth -= 1

File: geekbang/algorithm/source-code/2/lc111 二叉树的最小深度.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {
        // 不同于maxDepth的另一种写法
        // 也可以采用maxDepth的写法
        recur(root);
        return ans;
    }

    void recur(TreeNode node) {
        if (node == null) {
            ans = Math.min(ans, depth);
            return;
        }
        if (node.left == null && node.right == null) {
            ans = Math.min(ans, depth + 1);
            return;
        }
        depth++;
        if (node.left != null) recur(node.left);
        if (node.right != null) recur(node.right);
        depth--;
    }

    int depth = 0;
    int ans = 1 << 30;
}